packages/tokenbank_v1/contracts/impl/BaseWallet.sol

```diff
@@ -0,0 +1,125 @@
+
+
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "../iface/Wallet.sol";
+import "../iface/Module.sol";
+
+
+contract BaseWallet is Wallet
+{
+    event Received(
+        address indexed sender,
+        uint    value,
+        bytes   data
+    );
+
+    modifier onlyOwner
+    {
+        require(msg.sender == owner, "NOT_CALLED_BY_OWNER");
+        _;
+    }
+
+    modifier onlyModule
+    {
+        require(moduleIdx[msg.sender] != 0, "NOT_CALLED_BY_MODULE");
+        _;
+    }
+
+    function initialize(
+        address            _owner,
+        address[] calldata _modules
+        )
+        external
+    {
+        // TODO
+    }
+
+    function addModule(address _module)
+        external
+        onlyModule
+    {
+        // TODO
+    }
+
+    function delModule(address _module)
+        external
+        onlyModule
+    {
+        // TODO
+    }
+
+    function addFunction(bytes4 method, address module)
+        external
+        onlyModule
+    {
+        // TODO
+    }
+
+    function delFunction(bytes4 method)
+        external
+        onlyModule
+    {
+        // TODO
+    }
+
+    function transact(
+        address          to,
+        uint             value,
+        bytes   calldata data
+        )
+        external
+        onlyModule
+        returns (bytes memory result)
+    {
+        bool success;
+        // solium-disable-next-line security/no-call-value
+        (success, result) = to.call.value(value)(data);
+        if(!success) {
+            assembly {
+                returndatacopy(0, 0, returndatasize)
+                revert(0, returndatasize)
+            }
+        }
+        emit Transacted(msg.sender, to, value, data);
+    }
+
+    function()
+        external
+        payable
+    {
+        address module = msg.data.length == 0 ? address(0) : functions[msg.sig];
+        if(module == address(0)) {
+            if (msg.value > 0) {
+                emit Received(msg.sender, msg.value, msg.data);
+            }
+            return;
+        }
+
+        require(moduleIdx[module] !=0, "MODULE_UNAUTHORIZED");
+        assembly {
+            calldatacopy(0, 0, calldatasize())
+            let result := staticcall(gas, module, 0, calldatasize(), 0, 0)
```

Comment:
This is probably some temporary code (`staticcall` can't do any state changes), but will this be `delegatecall` or just `call`?

Reply:
I copied the code from Argent and I also have the same questions. Take a look at their comment here:

https://github.com/dong77/argent-contracts/blob/34da49e6defded4cce9602ea30baa53ef99ae76a/contracts/wallet/BaseWallet.sol#L119

Maybe you can figure it out.

Reply:
Ah interesting! So this will forward all `view` function calls on the wallet to the modules which some other smart contract using the wallet may depend on. Because it's a `staticcall` there's isn't really any risk of doing this because it's impossible to do any state changes. Because it seems riskless to me, I wonder why those functions need to be enabled for this forwarding to happen. I assume it's difficult otherwise to know when to forward the call or not (e.g. just a simple ETH transfer into the wallet should not do anything, but `msg.data.length > 0` is true).

Reply:
Let's talk about sometime later. The whole idea of method registration in a wallet seems wired to me as all read-only methods can be invocated directly in modules.

Reply:
A function call done through `transact` originates from the wallet address (`msg.sender == wallet`), so if the dapp depends on `msg.sender` to implement something, it needs to go through the wallet first (the dapp doesn't know about any modules, it only knows the wallet's address). At least that's my understanding.

Reply:
That's right, Brecht. The `transact` method is the killing feature for the wallet to transact on any arbitrary contracts with msg.sender == the_wallet_address.